services/src/main/java/io/grpc/services/HealthServiceImpl.java

```diff
@@ -50,16 +60,63 @@ public void check(HealthCheckRequest request,
     }
   }
 
-  void setStatus(String service, ServingStatus status) {
-    statusMap.put(service, status);
+  @Override
+  public void watch(HealthCheckRequest request,
+      final StreamObserver<HealthCheckResponse> responseObserver) {
+    final String service = request.getService();
+    synchronized (watchLock) {
+      ServingStatus status = statusMap.get(service);
+      responseObserver.onNext(getResponseForWatch(status));
+      watchers.put(service, responseObserver);
+    }
+    Context.current().addListener(
+        new CancellationListener() {
+          @Override
+          // Called when the client has closed the stream
+          public void cancelled(Context context) {
+            synchronized (watchLock) {
+              watchers.remove(service, responseObserver);
```

Comment:
I don't think there is promise about the equality of observers.   An interceptor may pass in a StreamObserver that has equality based on call state.   

I think the multimap _values_  should be an identity hash set.  That would make removal O(1), and avoid equality.

Reply:
`HashMultimap` is O(1) removal, but the equality is a valid concern. Switch to identity hash set.